roboteam_ai/src/utilities/OffensiveCoach.cpp

```diff
@@ -0,0 +1,312 @@
+
+    return score;
+}
+
+void OffensiveCoach::calculateNewPositions() {
+    auto it = offensivePositions.begin();
+    while (it < offensivePositions.end()) {
+        it->score = calculatePositionScore(it->position);
+        for (auto& robotPosition : robotPositions) {
+            if ((it->position - robotPosition.second.position).length() < Constants::ATTACKER_DISTANCE()) {
+                offensivePositions.erase(it);
+                break;
+            }
+        }
+        it++;
+    }
+
+    roboteam_msgs::GeometryFieldSize field = Field::get_field();
+    double xStart = 0 + marginFromLines;
+    double xEnd = field.right_penalty_line.begin.x - marginFromLines;
+    double yStart = -field.field_width + marginFromLines;
+    double yEnd = field.field_width - marginFromLines;
+
+    int attempts = 30;
+
+    for (int attempt = 0; attempt < attempts; attempt++) {
+        OffensivePosition position;
+
+        double x = (xEnd - xStart) * ( (double)rand() / (double)RAND_MAX) + xStart;
+        double y = (yEnd - yStart) * ( (double)rand() / (double)RAND_MAX) + yStart;
+
+        position.position = {x, y};
+
+        if (!Field::pointIsInField(position.position) || Field::pointIsInDefenceArea(position.position, false)) {
+            continue;
+        }
+
+        bool tooClose = false;
+        for (auto &robotPosition : robotPositions) {
+            if ((position.position - robotPosition.second.position).length() < Constants::ATTACKER_DISTANCE()) {
+                tooClose = true;
+                continue;
+            }
+        }
+
+        if (tooClose) {
+            continue;
+        }
+
+        position.score =  calculatePositionScore({x, y});
+
+        if (offensivePositions.empty()) {
+            offensivePositions.push_back(position);
+            continue;
+        }
+
+        OffensivePosition bestPos = position;
+        auto it = offensivePositions.begin();
+        while (it < offensivePositions.end()) {
+            if ((position.position - it->position).length() < Constants::OFFENSIVE_POSITION_DISTANCE()) {//Constants::ATTACKER_DISTANCE()) {
+                if (it->score >= bestPos.score) {
+                    bestPos.position = it->position;
+                    bestPos.score = it->score;
+                }
+                offensivePositions.erase(it);
+            }
+            it++;
+        }
+        offensivePositions.push_back(bestPos);
+    }
+
+    std::sort(offensivePositions.begin(), offensivePositions.end(), compareByScore);
+    if (offensivePositions.size() > maxPositions) {
+        offensivePositions.erase(offensivePositions.begin() + maxPositions, offensivePositions.end());
+    }
+    drawOffensivePoints();
+}
+
+bool OffensiveCoach::compareByScore(OffensivePosition position1, OffensivePosition position2) {
+    return position1.score > position2.score;
+}
+
+Vector2 OffensiveCoach::calculatePositionForRobot(std::shared_ptr<roboteam_msgs::WorldRobot> robot) {
```

Comment:
this method is too complex. please refactor

Reply:
Done